perun-dispatcher/src/test/java/cz/metacentrum/perun/dispatcher/unit/TaskSchedulerTest.java

```diff
@@ -47,4 +81,134 @@ public void sendToEngineTest() {
+		simpleSpy.setTask(simpleFutureTask);
+		simpleSpy.setSchedulingPool(schedulingPool);
+		Long timeLimit = 100L;
+		Task[] tasks = simpleSetup(timeLimit, schedulingPool);
+		Task testTask1 = tasks[0], testTask2 = tasks[1];
+		testTask2.setSourceUpdated(true);
+		schedulingPool.scheduleTask(testTask2, 4);
+		schedulingPool.scheduleTask(testTask1, 4);
+		Thread.sleep(timeLimit / 100);
+		schedulingPool.scheduleTask(testTask1, 4);
+		Thread.sleep((timeLimit / 100) * 8);
+		schedulingPool.scheduleTask(testTask1, 4);
+
+		simpleFutureTask.run();
+		assertFalse(simpleSpy.testFailed);
+	}
+
+	@Test
+	public void sourceUpdatedRecurrenceRunTest() throws InterruptedException, ExecutionException, TimeoutException {
+		//SchedulingPool schedulingPool = new SchedulingPoolRecurrenceSpy();
+		simpleSpy.setTask(recurrenceFutureTask);
+		recurrenceSpy.setSchedulingPool(schedulingPool);
+		Long timeLimit = 100L;
+		Task[] tasks = simpleSetup(timeLimit, schedulingPool);
+		Task testTask1 = tasks[0];
+		schedulingPool.scheduleTask(testTask1, 2);
+
+		recurrenceFutureTask.run();
+		assertFalse(recurrenceSpy.testFailed);
+	}
+
+	private Task[] simpleSetup(Long timeLimit, SchedulingPool schedulingPool) {
+		Properties properties = ((SchedulingPoolImpl)schedulingPool).getDispatcherProperties();
+		if(properties == null) {
+			properties = dispatcherProperties;
+		}
+		properties.setProperty("dispatcher.task.delay.time", timeLimit.toString());
+		properties.setProperty("dispatcher.task.delay.count", "1");
+		((SchedulingPoolImpl)schedulingPool).setDispatcherProperties(properties);
+		Task testTask1 = new Task();
+		testTask1.setService(service1);
+		testTask1.setFacility(facility1);
+		testTask1.setId(0);
+		testTask1.setStatus(TaskStatus.WAITING);
+		Task testTask2 = new Task();
+		testTask2.setId(2);
+		testTask2.setService(service2);
+		testTask2.setFacility(facility1);
+		testTask2.setStatus(TaskStatus.WAITING);
+		
+		return new Task[]{testTask1, testTask2};
+	}
+
+	private abstract class AbstractTaskSchedulerSpy extends TaskScheduler {
+		@Override
+		protected void initPerunSession() throws InternalErrorException {
+		}
+	}
+
+	/* WHY????
+	private class SchedulingPoolRecurrenceSpy extends SchedulingPoolImpl {
+		@Override
+		public void scheduleTask(Task task, int delayCount) {
+			super.scheduleTask(task, delayCount);
+		}
+	}
+	*/
+
+	private class SimpleTaskSchedulerSpy extends AbstractTaskSchedulerSpy {
+		int scheduledCounter = 0;
+		int scheduleLimit;
+		boolean testFailed = true;
+		FutureTask task;
+
+		public SimpleTaskSchedulerSpy(int scheduleLimit) {
+			this.scheduleLimit = scheduleLimit;
+		}
+
+		public void setTask(FutureTask task) {
+			this.task = task;
+		}
+
+		@Override
+		protected TaskScheduled sendToEngine(Task task) {
+			if (task.getId() == 2) {
+				testFailed = true;
+				this.task.cancel(true);
+				return TaskScheduled.ERROR;
+			}
+			scheduledCounter += 1;
+			if (scheduledCounter > scheduleLimit) {
```

Comment:
Switch condition to ``>=`` and inside IF call ``stop()`` so the TaskScheduler thread (represented here by ``SimpleTaskSchedulerSpy``) is not stuck and waiting for ``getWaitingTaskSchedule()`` inside its ``run()`` method.
And please make understandable comment in code so we know why it's like this.